source/agora/utils/Test.d

```diff
@@ -385,6 +385,42 @@ public struct TxBuilder
         return this;
     }
 
+    /***************************************************************************
+
+        Attaches to a range of tuples.
+        The tuple should have an `Output` and a `Hash` as its elements.
+
+        Params:
+            rng = the range of tuple
+
+        Returns:
+            A reference to `this` to allow for chaining
+
+    ***************************************************************************/
+
+    public ref typeof(this) attach (RNG) (scope RNG rng)
+        @safe pure nothrow return
+    {
+        alias ET = ElementType!RNG;
+
+        static assert(isInputRange!RNG);
+        static assert(ET.length == 2, "Parameter must be a tuple of 2 elements");
+        static assert(is(typeof(ET[0]) : const Output),
+            "Tuple's first element should be convertible to `Output`, not: `" ~
+            typeof(ET[0]).stringof ~ "`");
+        static assert(is(typeof(ET[1]) : const Hash),
+            "Tuple's second element should be convertible to `Hash`, not: `" ~
+            typeof(ET[1]).stringof ~ "`");
+
+        rng.each!((tup)
+        {
+            this.inputs ~= OutputRef(tup[0], tup[1]);
+            this.leftover.value.mustAdd(tup[0].value);
+        });
```

Comment:
If you can reuse code, do it. In this case:
```suggestion
        rng.each!((tup) { this.attach(tup); });
```

Also this has an interesting side-effect: The user could actually provide a `Transaction / index`, and if it wasn't for the static assert, it would work. But let's not loosen the static assert for now.

Reply:
Does `attach (in Output utxo, in Hash hash)` automatically recognize the `tup` elements and assign them respectively?

If not, then I think it is better to assign manually:
```D
rng.each!((tup) { this.attach(tup[0], tup[1]); });
```

Reply:
> If not, then I think it is better to assign manually:

It does. Tuples get expanded and then parameter matching is done.